### PR TITLE
8302635: Race condition in HttpBodySubscriberWrapper when cancelling request

### PR DESCRIPTION
**Pull request by dfuch**
The HttpBodySubscriberWrapper is a class that ensures that a subscriber will be subscribed to before it is completed. It also provides hooks to its two subclasses (one for HTTP/1, one for HTTP/2) that allows subclasses to register the susbscriber with the HttpClient at subscription time, and to unregister it when it is eventualy completed, or when the subscription is cancelled.

There is however a race condition that can happen when a subscription is cancelled: it can happen that unregister is called before register. The CancelRequestTest has been observed failing once or twice on personal jobs. Though the particular mechanics of this race is hard to understand, the logs of the tests have brought sufficient evidence that this is what was happening. 

The symptom is finding one subscriber still registered after completion of the exchange:

test CancelRequestTest.testGetSendAsync("https://localhost:42711/https1/x/same/interrupt", true, true): failure
java.lang.AssertionError: WARNING: tracker for HttpClientImpl(13) has outstanding operations:
Pending HTTP Requests: 0
Pending HTTP/1.1 operations: 0
Pending HTTP/2 streams: 0
Pending WebSocket operations: 0
Pending TCP connections: 0
Pending Subscribers: 1
Total pending operations: 0
Facade referenced: true
Selector alive: true

The proposed fix hoist special hooks for register/unregister in the superclass, merges all various volatile boolean states into a single int state, and protect the state changes  to subscribed/register/unregister by the same subscription lock.
If cancelling the subscription happens at around the same time that the subscriber is subscribed this ensures that the subscriber won't be removed from the map before it is added.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302635](https://bugs.openjdk.org/browse/JDK-8302635): Race condition in HttpBodySubscriberWrapper when cancelling request

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [47887520](https://git.openjdk.org/jdk/pull/12587/files/47887520cfe1590b6a8cb3bf16173c2a92abe619)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12587/head:pull/12587` \
`$ git checkout pull/12587`

Update a local copy of the PR: \
`$ git checkout pull/12587` \
`$ git pull https://git.openjdk.org/jdk pull/12587/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12587`

View PR using the GUI difftool: \
`$ git pr show -t 12587`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12587.diff">https://git.openjdk.org/jdk/pull/12587.diff</a>

</details>

### CODE REVIEW

File: src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java [hunk 1]
```diff
@@ -58,11 +58,16 @@ public void request(long n) { }
...
+    static final int COMPLETED = 4;
+    static final int CANCELLED = 8;
+    static final int UNREGISTERED = 16;
+
     static final AtomicLong IDS = new AtomicLong();
     final long id = IDS.incrementAndGet();
     final BodySubscriber<T> userSubscriber;
-    final AtomicBoolean completed = new AtomicBoolean();
-    final AtomicBoolean subscribed = new AtomicBoolean();
+    volatile int state;
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:70]: Hello Daniel, should we make this `private` and all the newly introduced state values, too? Or maybe you want to leave it package private to be consistent with some other fields here?

File: src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java [hunk 2]
```diff
@@ -125,34 +172,139 @@ private void propagateError(Throwable t) {
         }
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:172]: I haven't fully wrapped my head around the possible flow of the user subscription, but is there a possiblity where this call to `onError(...)` here results in an reentrant call to this `propagateError()`? For that matter, not just reentrant but perhaps from a different thread concurrently? The reason I ask is, should we call this onError just once? The `java.util.concurrent.Flow.Subscriber.onError(Throwable)` method says this:
> Method invoked upon an unrecoverable error encountered by a
> Publisher or Subscription, after which no other Subscriber
> methods are invoked by the Subscription. 

So I'm wondering if we should maintain some state to only invoke it once?

File: src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java (hunk 1 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:70]: Good idea.

Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:70]:
```suggestion
    private volatile int state;
```

File: src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java (hunk 2 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:172]: This is the purpose of this class: complete() and propagateError() should ensure that onError() is only called once in the wrapped subscriber. The markCompleted() call should ensure that, even if there is a reentrant call.

Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:172]: You are right - the error propagation through `propagateError()` always happens through the `complete()` method of this wrapper class and that method has the necessary state management to call this only once.

**Review by jaikiran — APPROVED** [review]
The changes look good to me. `HttpBodySubscriberWrapper` and `Http1Exchange` files will need a copyright year update, before integrating.

**Pull request closed without merging**